### PR TITLE
added symbolic sum to valexpr

### PR DESCRIPTION
**Pull request by pjljvandelaar**
Copied + modified from 1897 branch in subversion
To do: also copy all tests

### CODE REVIEW

**Comment by pjljvandelaar** [comment]
Two hlint warnings are not removed.
1. To cover Eq (in deriving cause) /= must be covered.
    Test must return true, easiest case is thus not (s /= s)
2. Extension of tests -> with return it is clearer how to extend.
One TODO
1. Must still be investigated!

File: sys/defs/src/Sum.hs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+--------------------------------------------------------------------}
+-- | /O(log n)/. Add a term to a sum.
+add :: Ord a => a -> Sum a -> Sum a
+add t = addMultiply t 1
+
+-- | /O(log n)/. Subtract a term from a sum.
+subtract :: Ord a => a -> Sum a -> Sum a
+subtract t = addMultiply t (-1)
+
+-- | /O(log n)/. Add a term multiplied by the given coefficient to a sum.
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/Sum.hs]: subtract t needs an additional parameter!

File: sys/defs/src/Sum.hs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+addMultiply :: Ord a => a -> Integer -> Sum a -> Sum a
+addMultiply _ 0 s = s                                    -- invariant: no term with multiplier 0 is stored.
+addMultiply x m s = (Sum . Map.alter increment x . unSum) s
+    where
+        increment :: Maybe Integer -> Maybe Integer
+        increment Nothing            = Just m
+        increment (Just n) | n == -m = Nothing           -- Terms with multiplier zero are removed
+        increment (Just n)           = Just (n+m)
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/Sum.hs]: No:
3*x + 4*x = (3+4)*x

File: sys/defs/src/Sum.hs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+--------------------------------------------------------------------}
+-- | /O(1)/. The number of distinct terms in the sum.
+nrofTerms :: Sum a -> Int
+nrofTerms = Map.size . unSum
+
+{--------------------------------------------------------------------
+  Sums and multiplications
+--------------------------------------------------------------------}
+-- | /O(log n)/. Add a term to a sum.
+add :: Ord a => a -> Sum a -> Sum a
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/Sum.hs]: Is not a number: is a symbolic sum

File: sys/defs/src/Sum.hs [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+sums = foldlStrict sum (Sum Map.empty)
+
+-- | /O(n+m)/. The sum of two sums.
+--
+-- The implementation uses the efficient /hedge-union/ algorithm.
+-- Hedge-union is more efficient on (bigsum `sum` smallsum).
+sum :: Ord a => Sum a -> Sum a -> Sum a
+sum (Sum m1) (Sum m2) = Sum $ Map.filter (0/=) $ Map.unionWith (+) m1 m2
+
+-- | /O(n)/. Multiply the sum /ms/ with the constant /x/.
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/Sum.hs]: removed & swapped

File: sys/defs/src/Sum.hs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+multiply :: Integer -> Sum a -> Sum a
+multiply 0 _ = Sum Map.empty
+multiply n s = (Sum . Map.map (n *) . unSum) s
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/Sum.hs]: What is the advance of being a Functor?

File: sys/defs/src/Sum.hs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+toMultiplierList :: Sum a -> [(a, Integer)]
+toMultiplierList = toDistinctAscMultiplierList
+
+-- | /O(n)/. Convert the sum to a distinct ascending list of term\/multiplier pairs.
+toDistinctAscMultiplierList :: Sum a -> [(a, Integer)]
+toDistinctAscMultiplierList = Map.toAscList . unSum
+
+-- | /O(n*log n)/. Create a sum from a list of term\/multiplier pairs.
+fromMultiplierList :: Ord a => [(a, Integer)] -> Sum a 
+fromMultiplierList = Sum . Map.filter (0/=) . Map.fromListWith (+)
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/Sum.hs]: occurs only twice, so in doubt....

File: sys/defs/src/Sum.hs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+-- each term appears only once.
+-- /The precondition (input list is strictly ascending) is not checked./
+fromDistinctAscMultiplierList :: [(a, Integer)] -> Sum a 
+fromDistinctAscMultiplierList = Sum . Map.filter (0/=) . Map.fromDistinctAscList
+
+{--------------------------------------------------------------------
+  Utilities
+--------------------------------------------------------------------}
+
+-- TODO : Use foldl' from base?
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/Sum.hs]: done

File: sys/defs/src/TxsDefs.hs [hunk 8]
```diff
@@ -110,7 +93,7 @@ import           StatId
 import           VarId
 
 import           ValExprDefs
-import           ValExprImpls
+import           ValExprImpls   as X
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/TxsDefs.hs]: feel free to change, I have no idea for a better name

File: sys/defs/src/TxsShow.hs [hunk 9]
```diff
@@ -269,6 +270,18 @@ instance PShow v => PShow (ValExpr v) where
       =  "(" ++ pshow t ++ " / " ++ pshow n ++ " )"
     pshow (view -> Vmodulo t n)
       =  "(" ++ pshow t ++ " % " ++ pshow n ++ " )"
+    pshow (view -> Vsum s)
+      = showList (Sum.toMultiplierList s)  
+      where
+        showList []     = "0"
+        showList [x]    = showElem x
+        showList (x:xs) = "( " ++ showElem x ++ " + " ++ showList xs ++ " )"
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/TxsShow.hs]: Yes

 5 *( 4 + 3)  <> 5*4 + 3

File: sys/defs/src/ValExprImpls.hs [hunk 10]
```diff
@@ -181,6 +189,57 @@ cstrModulo vet ven@(view -> Vconst (Cint n)) | n == 0 = Trace.trace "Error in mo
 cstrModulo (view -> Vconst (Cint t)) (view -> Vconst (Cint n)) = cstrConst (Cint (t `mod` n) )
 cstrModulo vet ven = ValExpr (Vmodulo vet ven)
 
+-- | Apply operator Minus on the provided value expression.
+-- Preconditions are /not/ checked.
+cstrMinus :: Ord v => ValExpr v -> ValExpr v
+cstrMinus (view -> Vconst (Cint x)) = cstrConst (Cint (-x))
+cstrMinus (view -> Vsum s)          = cstrSum (Sum.multiply (-1) s)
+cstrMinus v                         = ValExpr (Vsum (Sum.fromDistinctAscMultiplierList [(v,-1)]))
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/ValExprImpls.hs]: fromDistinctAscMultiplierList is faster than the alternative (in general, maybe for length 1 doesn't matter)

File: sys/defs/src/ValExprImpls.hs [hunk 11]
```diff
@@ -181,6 +189,57 @@ cstrModulo vet ven@(view -> Vconst (Cint n)) | n == 0 = Trace.trace "Error in mo
...
+-- | Apply operator sum on the provided sum of value expressions.
+-- Preconditions are /not/ checked.
+cstrSum :: Ord v => Sum (ValExpr v) -> ValExpr v
+cstrSum ms = 
+    let (adds, nonadds) = Sum.partition isSum ms in
+            cstrSum' $ foldl Sum.sum nonadds (Sum.foldMultiplier toSumList [] adds)
+    where                                                        
+        toSumList :: ValExpr v -> Integer -> [Sum (ValExpr v)] -> [Sum (ValExpr v)]
+        toSumList (view -> Vsum s) n  = (:) (Sum.multiply n s)
+        toSumList _                _  = error "ValExprImpls.hs - cstrSum - Unexpected ValExpr"
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/src/ValExprImpls.hs]: due to the partitioning based isSum function, I know the alternative can NEVER occur.
However, hlint or Wall require that all cases are handled -> for these cases error is more than acceptable!

File: sys/defs/test/GenSum.hs [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+import Test.QuickCheck.Gen(chooseAny)
+
+import Sum
+
+newtype GenSum a = GenSum (Sum a)
+    deriving (Show)
+    
+instance (Ord a, Arbitrary a) => Arbitrary (GenSum a) where
+  arbitrary = do
+    i <- choose (0,100)
```
[the remark this answers is not in the document]
Comment (pjljvandelaar) [reply sys/defs/test/GenSum.hs]: Changed: but now I sometimes get:

Preprocessing library behave-0.2.1.0...
Preprocessing test suite 'behave-test' for behave-0.2.1.0...
[ 8 of 11] Compiling TestHelperFuncContent ( test\TestHelperFuncContent.hs, .stack-work\dist\1f7101f2\build\behave-test\behave-test-tmp\TestHelperFuncContent.o )

C:\TorXakisSandbox\TorXakis.git\sys\behave\test\TestHelperFuncContent.hs:111:1: warning:
    Pattern match checker exceeded (2000000) iterations in
    an equation for `identicalVExpr'. (Use -fmax-pmcheck-iterations=n
    to set the maximun number of iterations to n)
Linking .stack-work\dist\1f7101f2\build\behave-test\behave-test.exe ...
Installing library in
C:\TorXakisSandbox\TorXakis.git\.stack-work\install\84603a22\lib\x86_64-windows-ghc-8.0.2\behave-0.2.1.0-BNPO91JEA9gE6dtLBVVwox
Registering behave-0.2.1.0...

File: sys/defs/src/Sum.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (capitanbatata) [reply sys/defs/src/Sum.hs]: Three times to be precise ;)

File: sys/defs/test/GenSum.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (capitanbatata) [reply sys/defs/test/GenSum.hs]: For reference, see this question https://stackoverflow.com/questions/40879983/what-sort-of-iterations-does-ghc-do-while-compiling-patterns Maybe we need to report an issue to ghc.

**Comment by pjljvandelaar** [comment]
linked to #133

**Pull request merged**